Harja 1735 tarjousnakyma toimenkuva bugit

rckopteri opened the pull request:
Tarjousnäkymän toimenkuvan uusi toimenkuva meni väärään kohtaan listassa, kun lisäsi uuden. Ei kunnioittanut :jarjestys avaimen arvoa.
Kovakoodatut hoitovuoden avaimet vuotivat myös toimenkuvan tallennukseen asti, joka johti siihen, että toimenkuvalle tallennettiin väärälle hoitovuodelle arvoja.
Samalla estettiin "lisää" napin näyttäminen, jos valittavia toimenkuvia ei ole enää saatavilla.

rckopteri committed:
Korjaa lisätyn toimenkuvan tallennus ja järjestys käyttöliittymässä

rckopteri committed:
Merge branch 'develop' into HARJA-1735-tarjousnakyma-toimenkuva

rckopteri committed:
Estä uuden toimenkuvan lisääminen, jos niitä ei ole jäljellä lisättäväksi asti.

rckopteri committed:
Merge branch 'develop' into HARJA-1735-tarjousnakyma-toimenkuva

samnes commented on src/clj/harja/kyselyt/tarjous_kyselyt.clj:
```diff
@@ -302,11 +302,13 @@
-                                                                                                                    :urakkakohtainen-nimi (:nimi toimenkuva)})))
+                                                                (let [toimenkuva-kannasta (first (toimenkuva-kyselyt/hae-urakan-toimenkuva db {:toimenkuva (:toimenkuva toimenkuva)
+                                                                                                                                               :urakkaid urakka-id}))]
+                                                                  (if-not (nil? toimenkuva-kannasta)
```
Onko tässä tarkoituksella logiikka "jos toimenkuva löytyy kannasta, lisätään uusi, muuten palautetaan vanha"?

rckopteri replied:
Ei ole. Toimenkuvat pyöritetään tuon läpi vuosittain koko hoitokauden kaikille hoitovuosille. Tarkoituksena on ensimmäisenä hoitovuonna lisätä toimenkuva johto_ja_hallintokorvaus_toimenkuva -tauluun urakan id:llä ja muina vuosina vain hyödyntää sieltä tullut toimenkuvan id:tä.

Tämä bugi aiheutti sen, että urakalle ei lisätty uutta toimenkuvaa, vaan se kierrätti niitä vanhoja, jotka on jo aiemmin kovakoodattu sinne tauluun kaikille käyttöön, vaikkakin ne ei ole kaikkien käytössä. (älä kysy miksi, kun kukaan ei tiedä)

Joka ei aiheuttanut ongelmia tallennusvaiheessa, mutta poistamisesa, koska se käytetty toimenkuva ei ollut yksiselitteisesti urakkakohtainen.

Erittäin hyvä bogaus!.

samnes commented on src/cljs/harja/views/urakka/suunnittelu/tarjous_kustannussuunnitelma/tarjous_nakyma.cljs:
```diff
@@ -1,21 +1,22 @@
+    [tuck.core :as tuck]))
 
 (defonce virheet-atom (atom {}))
+(defonce uusi-toimenkuva-valittavana (atom false))
```
Näkymässä siirrytty suurelta osin tuck-staten käyttöön joten atomin käyttö on siinä mielessä huonompi valinta että ei pysty näkymän tilasta helposti debugaamaan ja seuraamaan.

rckopteri replied:
Vaihdoin käyttämään app-statea ja tuckkia

samnes reviewed (commented):
Tuo backend logiikka olisi ainakin hyvä varmistaa ennen mergeä.

rckopteri committed:
Merge branch 'develop' into HARJA-1735-tarjousnakyma-toimenkuva

rckopteri committed:
Korjaa uuden toimenkuvan lisäys urakalle

rckopteri committed:
Vaihda atom app-state attribuuttiin

solita-staskila commented on src/cljs/harja/views/urakka/suunnittelu/tarjous_kustannussuunnitelma/tarjous_nakyma.cljs:
```diff
@@ -108,8 +112,10 @@
+      :voi-lisata? voi-lisata?
+      :uusi-rivi (fn [rivi]
+                   (e! (tarjous-tiedot/->ToggleUusiToimenkuvaValittavana true))
+                   (merge (assoc rivi :id -1 :nimi "" :jarjestys 99 :yhteensa 0) vuosi-map))
```
Nää pomppii käyttöliittymässä vähän lisäysten jälkeen myös, onks toi ihan normaalia? 

Tossa halutaan ilmeisesti noudattaa tiettyä järjestystä?


[Screencast from 2025-10-16 13-26-32.webm](https://github.com/user-attachments/assets/09c30a14-8462-4c55-bc52-40d21f2438a8)


Eli tuossa 6s kohdalla, sen jälkeen kun lisännyt Apulaisen, muokatessa vaihtaa vielä järjestystä.

Tuon jälkeen näyttää kuitenkin pysyvän, eli ei nyt ole niin hirveää..

solita-staskila replied:
Kuitenkin haluttaisiin välttyä tällaiselta pomppimiselta käyttöliittymässä

rckopteri replied:
Kokeilin -25 ja -24 alkaville urakoille tuota. En saa toistumaan. Juurikin halutaan pitää ne toimenkuvat samassa järjestyksessä siihen asti, kunnes ne on tallennettu. Tallennuksen jälkeen ne menee siihen järjestykseen, mihin ne on tietokannassa määritelty.

rckopteri replied:
Gridille tosiaan annettaan 
:jarjesta :jarjestys 
Määritys, joten sen pitäisi noudattaa aina tuota: :jarjestys 99 arvoa. Eli olla viimeisenä.

rckopteri replied:
Oiskohan tuossa se, että oot lisännyt sinne niin monta, että tuo järjestys 99 numero ei riitä. Sen pitäisi olla indeksipohjainen. Se saattaa siis joskus laittaa 99 arvoiset eri järkkään, kun niitä lisää monta kerralla.

solita-staskila replied:
Iin Mhu taisi olla valittuna, X(?) hoitovuosi,  ne pomppivat myös tuossa lisätessä (videollakin näkyy kun lisään tuon apulaisen). 

Tarviiko katsoa uudelleen?

rckopteri replied:
Sain tämän toistumaan -21 alkavalla urakalla. Jännä, että noilla muilla ei tullut tätä vastaan. Lisään tuohon indeksipohjaisuutta, niin vältetään tämä noillakin urakoilla, joita ei vielä tueta.

rckopteri committed:
Hae ohjauskavalla gridin muokkaustila ja päättele toimenkuvien määrä

jota hyödynnetään järjestysindeksissä, ettei toimenkuvat mene väärään järjestykseen.

rckopteri committed:
Merge branch 'develop' into HARJA-1735-tarjousnakyma-toimenkuva
